src/Network/Ethereum/Web3/Provider.hs

```diff
@@ -35,18 +35,14 @@ import           Network.HTTP.Client.TLS    (tlsManagerSettings)
 import           Network.HTTP.Client        (defaultManagerSettings)
 #endif
 
-import           Network.JsonRpc.TinyClient (Remote, ServerUri)
-
 -- | Any communication with Ethereum node wrapped with 'Web3' monad
-newtype Web3 a = Web3 { unWeb3 :: ReaderT (ServerUri, Manager) IO a }
+newtype Web3 a = Web3 { unWeb3 :: ReaderT (Provider, Manager) IO a }
     deriving (Functor, Applicative, Monad, MonadIO, MonadThrow)
 
-instance MonadReader (ServerUri, Manager) Web3 where
+instance MonadReader (Provider, Manager) Web3 where
```

Comment:
`instance MonadReader (ServerUri, Manager) Web3` is minimal JSON-RPC client requirements, also it can be created for new `newtype Web3 a = Web3 { unWeb3 :: ReaderT (Provider, Manager) IO a }` type too. 

```haskell
instance MonadReader (ServerUri, Manager) Web3 where
     ask = Web3 (\((Provider (HttpProvider a) _), b) -> (a, b) <$> ask)
     local f = Web3 . local f . unWeb
```

Reply:
I can't have two instances of `MonadReader` for the new `Web3` type. I'd get an error indicating conflicting functional dependencies. I need the `Web3` to have access to the full `Provider` in context because I need access to the signing config in the `sendTx` function of `Method.hs`: `signingConfigM <- asks (signingConfiguration . fst)`. Is there a clean way to resolve this without defining something like an additional `Web3Light` type?

Your remaining outstanding comments are all related to this change. `TinyClient.hs` currently needs to import `Provider.hs`, and not the other way around, because it needs to import the `Provider` type.

Reply:
Yup, I see the problem, ok, I need some time to resolve it.